packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -48,43 +50,101 @@ export class AssetStaging extends Construct {
   public readonly sourcePath: string;
 
   /**
-   * A cryptographic hash of the source document(s).
+   * A cryptographic hash of the asset.
+   *
+   * @deprecated see `assetHash`.
    */
   public readonly sourceHash: string;
 
+  /**
+   * A cryptographic hash of the asset.
+   */
+  public readonly assetHash: string;
+
   private readonly fingerprintOptions: FingerprintOptions;
 
   private readonly relativePath?: string;
 
+  private readonly bundleDir?: string;
+
   constructor(scope: Construct, id: string, props: AssetStagingProps) {
     super(scope, id);
 
+    if (props.assetHash && props.assetHashType && props.assetHashType !== AssetHashType.CUSTOM) {
+      throw new Error(`Cannot specify \`${props.assetHashType}\` for \`assetHashType\` when \`assetHash\` is specified. Use \`CUSTOM\` or leave \`undefined\`.`);
+    }
+
     this.sourcePath = props.sourcePath;
     this.fingerprintOptions = props;
-    this.sourceHash = FileSystem.fingerprint(this.sourcePath, props);
+
+    if (props.bundling) {
+      this.bundleDir = this.bundle(props.bundling);
+    }
+
+    const hashCalculation = props.assetHash ? AssetHashType.CUSTOM : props.assetHashType ?? AssetHashType.SOURCE;
+    if (hashCalculation === AssetHashType.SOURCE) {
+      this.assetHash = this.fingerprint(props.sourcePath);
+    } else if (hashCalculation === AssetHashType.BUNDLE) {
+      if (!this.bundleDir) {
+        throw new Error('Cannot use `AssetHashType.BUNDLE` when `bundling` is not specified.');
+      }
+      this.assetHash = this.fingerprint(this.bundleDir);
+    } else if (hashCalculation === AssetHashType.CUSTOM) {
+      if (!props.assetHash) {
+        throw new Error('`assetHash` must be specified when `assetHashType` is set to `AssetHashType.CUSTOM`.');
+      }
+      this.assetHash = props.assetHash;
+    } else {
+      throw new Error('Unknown asset hash type.');
+    }
 
     const stagingDisabled = this.node.tryGetContext(cxapi.DISABLE_ASSET_STAGING_CONTEXT);
-    if (stagingDisabled) {
+    if (stagingDisabled && this.bundleDir) {
+      this.stagedPath = this.bundleDir;
+    } else if (stagingDisabled) {
       this.stagedPath = this.sourcePath;
     } else {
-      this.relativePath = 'asset.' + this.sourceHash + path.extname(this.sourcePath);
-      this.stagedPath = this.relativePath; // always relative to outdir
+      this.relativePath = `asset.${this.assetHash}${path.extname(this.bundleDir ?? this.sourcePath)}`;
```

Comment:
So this means that if we bundle and use a SOURCE hash type, the relative directory will be based on the source hash and not the bundle. I am okay with that.

Reply:
Yes, I changed my mind on this. It's easier.

Reply:
Yass!